std/experimental/allocator/building_blocks/ascending_page_allocator.d

```diff
@@ -0,0 +1,418 @@
+
+        // Give memory back just by changing page protection
+        version(Posix)
+        {
+            import core.sys.posix.sys.mman : mmap, mprotect,
+                   MAP_PRIVATE, MAP_ANON, MAP_FAILED, PROT_NONE, PROT_WRITE, PROT_READ;
+            auto ret = mprotect(result, goodSize, PROT_WRITE | PROT_READ);
+            enforce(ret == 0, "Failed to allocate memory, mprotect failure");
+        }
+        else version(Windows)
+        {
+            import core.sys.windows.windows : VirtualProtect, PAGE_READWRITE;
+            uint oldProtect;
+            auto ret = VirtualProtect(result, goodSize, PAGE_READWRITE, &oldProtect);
+            enforce(ret != 0, "Failed to allocate memory, VirtualProtect failure");
+        }
+
+        return cast(void[]) result[0 .. n];
+    }
+
+    /**
+    Rounds the requested size to the next multiple of the page size
+    */
+    size_t goodAllocSize(size_t n)
+    {
+        return n.roundUpToMultipleOf(pageSize);
+    }
+
+    /**
+    Removes the read/write protection of the pages which the passed buffer covers.
+    Hints the OS to reclaim the physical memory(``madvise`` for Posix and
+    ``VirtualUnlock`` for Windows).
+    If the allocator is no longer valid, and all objects have been deallocated,
+    the range of virtual addresses is unmapped.
+    */
+    version(Posix)
+    {
+        bool deallocate(void[] buf)
+        {
+            import core.sys.posix.sys.mman : posix_madvise, POSIX_MADV_DONTNEED, mprotect, PROT_NONE, munmap;
+            import std.exception : enforce;
+
+            size_t goodSize = goodAllocSize(buf.length);
+            auto ret = mprotect(buf.ptr, goodSize, PROT_NONE);
+            enforce(ret == 0, "Failed to deallocate memory, mprotect failure");
+
+            // madvise to let the OS reclaim the resources
+            ret = posix_madvise(buf.ptr, goodSize, POSIX_MADV_DONTNEED);
+            enforce(ret == 0, "Failed to deallocate, posix_madvise failure");
+            pagesUsed -= goodSize / pageSize;
+
+            if (!valid && pagesUsed == 0)
+            {
+                munmap(data, numPages * pageSize);
+                data = null;
+            }
+
+            return true;
+        }
+    }
+    else version(Windows)
+    {
+        bool deallocate(void[] buf)
+        {
+            import core.sys.windows.windows : VirtualUnlock, VirtualProtect,
+                   VirtualFree, PAGE_NOACCESS, MEM_RELEASE;
+            import std.exception : enforce;
+
+            uint oldProtect;
+            size_t goodSize = goodAllocSize(buf.length);
+            auto ret = VirtualProtect(buf.ptr, goodSize, PAGE_NOACCESS, &oldProtect);
+            enforce(ret != 0, "Failed to deallocate memory, VirtualProtect failure");
+
+            // MSDN states for VirtualAlloc: Calling VirtualUnlock on a range
+            // of memory that is not locked releases the pages from the process's working set.
+            VirtualUnlock(buf.ptr, goodSize);
```

Comment:
missing check for success/failure.

Reply:
Also see MEM_RESET in VirtualAlloc, might be a better match for your needs. Anyhow - benchmark/test is appropriate.

Reply:
VirtualUnlock here is expected to fail, but have the intended behavior [1].
I'll also experiment in more detail with all params for VirtualAlloc.

[1] - https://msdn.microsoft.com/en-us/library/windows/desktop/aa366910(v=vs.85).aspx

Reply:
@jercaianu you may want to put this as a code comment just before the call

Reply:
This blog post from MSDN might come in handy: https://blogs.msdn.microsoft.com/oldnewthing/20170113-00/?p=95185

Reply:
@nemanja-boric-sociomantic 
Thanks for the post :)

After reading this I think that only VirtualUnlock is not enough, because as it states here, the pages would be evicted from the working set, but memory would still be accessible. From my understanding this means that the pages will be swapped out on disk, right?

Maybe we can use this together with VirtualAlloc/MEM_RESET, which states that it will make the contents unavailable after eviction.

I'll test this, but can someone please confirm/infirm, just in theory, if my judgement is right?
Thanks

Reply:
Yes, `VirtualUnlock` would do similar thing to the `munlock` in Posix - just allow system to swap pages out, but they would still be transferred to pagefile (which you don't need). I indeed think you want `VirtualAlloc/MEM_RESET`, I don't think you need to call `VirtualUnlock` at all.